e2e/libs/node/node.py

```diff
@@ -1,26 +1,121 @@
+        if int(index) < 0 or int(index) > len(cls.all_nodes):
+             raise Exception(f"invalid parameter: index={index}")
+
+        return cls.all_nodes[int(index)]['name']
+
+    @classmethod
+    def get_index_by_name(cls, node_name):
+        if node_name == "":
+            raise Exception(f"empty parameters: node_name={node_name}")
+
+        logging.info(f"getting node index by name {node_name}")
+        index = 0
+        for node in cls.all_nodes:
+            if node["name"] == node_name:
+                logging.debug(f"find the node name {node_name} index: {index}")
+                return index
+            index += 1
+
+        raise Exception(f"failed to get index by node name {node_name}")
+
+    @classmethod
+    def get_node_state(cls, node_name):
+        if node_name == "":
+            raise Exception(f"empty parameters: node_name={node_name}")
+
+        logging.info(f"getting node {node_name} state")
+        client = globalVars.K8S_API_CLIENT
+        node_status = client.read_node_status(node_name)
+        for node_cond in node_status.status.conditions:
+            logging.debug(f"node_cond.type:{node_cond.type}, node_cond.status:{node_cond.status}")
+            if node_cond.type == "Ready" and \
+                node_cond.status == "True":
+                 return node_cond.type
+        return "NotReady"
+
+    @classmethod
+    def refresh_node_list(cls):
+        logging.info("refreshing node list")
+
+        resp = globalVars.K8S_API_CLIENT.list_node()
+        if resp == "":
+            raise Exception("failed to get cluster nodes")
+
+        nodes = resp.items
+        if len(nodes) == 0:
+            raise Exception("failed to get cluster nodes: empty items")
+
+        for item in nodes:
+            worker_node = False
+            if ('node-role.kubernetes.io/control-plane' not in item.metadata.labels and \
+                'node-role.kubernetes.io/master' not in item.metadata.labels):
+                worker_node = True
+            elif ('node-role.kubernetes.io/worker' in item.metadata.labels and \
+                item.metadata.labels['node-role.kubernetes.io/worker'] == 'true'):
+                worker_node = True
+
+            if worker_node:
+                for address in item.status.addresses:
+                    if address.type == "ExternalIP":
+                        logging.debug(f'node name:{item.metadata.name}, ExternalIP:{address.address}')
+                        node_name = item.metadata.name
+                        ip_address = address.address
+                        break
+                    elif address.type == "InternalIP":
+                        logging.debug(f'node name:{item.metadata.name}, InternalIP:{address.address}')
+                        node_name = item.metadata.name
+                        ip_address = address.address
+                        break
+
+                # if node is exist then update the ip address or add new one
```

Comment:
What happens if the instance has both external and internal IPs? Do we randomly pick one?

Reply:
I understand the situation you mentioned, so the external IP address will be preferred.

Reply:
In `refresh_node_list()`, it will get each node information from the K8S API server, if the type is `ExternalIP` the IP address will be used first.
https://github.com/longhorn/longhorn-tests/blob/5290967ed6110b1c7dbb194db19be29810102519/e2e/libs/node/node.py#L97-L106

Reply:
If the `item.status.addresses` sequence is [<internal-addr>, <external-addr>], this logic will break `elif address.type == "InternalIP":`.. right? Is there a guarantee in the address sequence that it will start with external IP?

Reply:
After searching the k8s document and querying the information of the local cluster, it is true that the order of the address fields is not guaranteed.
![image](https://github.com/longhorn/longhorn-tests/assets/17548901/601d4ee8-e0cf-4380-a86f-370669c51a3d)

So I added a piece of logic so that the address fields must be arranged in the following order: `ExternalIP` -> `InternalIP` -> `Hostname`
https://github.com/longhorn/longhorn-tests/blob/3e1a74f90f624a537a1cabf9638425ddf3cc89ac/e2e/libs/node/node.py#L98-L102

And I confirmed that it can capture the correct IP address.
![image](https://github.com/longhorn/longhorn-tests/assets/17548901/b5cc7c69-c3d7-440b-8fd0-d528ef52170f)